File: rust/cudf-sys/cpp/contiguous_split.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION.
// SPDX-License-Identifier: Apache-2.0

#include "cudf-sys/src/contiguous_split.rs.h"
#include "cudf-sys/src/lib.rs.h"
#include "cudf-sys/helpers.hpp"

#include <cudf/contiguous_split.hpp>

namespace cudf_sys {

// -- PackedColumns --

rust::Vec<uint8_t> PackedColumns::metadata_to_host() const {
  rust::Vec<uint8_t> out;
  if (packed_.metadata) {
    auto const& meta = *packed_.metadata;
    out.reserve(meta.size());
    for (auto b : meta) {
      out.push_back(b);
    }
  }
  return out;
}

std::size_t PackedColumns::metadata_size() const {
  return packed_.metadata ? packed_.metadata->size() : 0;
}

std::size_t PackedColumns::gpu_data_size() const {
  return packed_.gpu_data ? packed_.gpu_data->size() : 0;
}

// -- PackedTableVec --

std::size_t PackedTableVec::size() const {
  return vec_.size();
}

std::unique_ptr<Table> PackedTableVec::unpack_at(std::size_t index) const {
  if (index >= vec_.size()) {
    throw std::out_of_range("PackedTableVec index out of range");
  }
  auto view = vec_[index].table;
  return std::make_unique<Table>(std::make_unique<cudf::table>(view));
}

// -- Free functions --

std::unique_ptr<PackedColumns> pack_table(Table const& tbl, std::size_t stream) {
  auto packed = cudf::pack(tbl.cached_view(), S(stream));
  return std::make_unique<PackedColumns>(std::move(packed));
}

std::size_t packed_size_of(Table const& tbl, std::size_t stream) {
  return cudf::packed_size(tbl.cached_view(), S(stream));
}

std::unique_ptr<Table> unpack_packed(PackedColumns const& packed) {
  auto view = cudf::unpack(packed.inner());
  return std::make_unique<Table>(std::make_unique<cudf::table>(view));
}

std::unique_ptr<PackedTableVec> contiguous_split_table(Table const& tbl, rust::Slice<int32_t const> splits, std::size_t stream) {
  std::vector<cudf::size_type> split_vec(splits.begin(), splits.end());
  auto result = cudf::contiguous_split(tbl.cached_view(), split_vec, S(stream));
  return std::make_unique<PackedTableVec>(std::move(result));
}

}  // namespace cudf_sys

File: rust/cudf-sys/src/contiguous_split.rs
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION.
// SPDX-License-Identifier: Apache-2.0

#[cxx::bridge(namespace = "cudf_sys")]
pub mod ffi {
    unsafe extern "C++" {
        include!("cudf-sys/lib.hpp");

        // Shared types from the core bridge.
        type Table = crate::ffi::Table;

        /// Opaque packed columns (host metadata + device data).
        type PackedColumns;

        /// Opaque vector of packed tables from `contiguous_split`.
        type PackedTableVec;

        // -- Pack / Unpack --

        /// Packs a table into a contiguous device buffer with host metadata.
        fn pack_table(tbl: &Table, stream: usize) -> Result<UniquePtr<PackedColumns>>;

        /// Returns the size in bytes that packing would require.
        fn packed_size_of(tbl: &Table, stream: usize) -> Result<usize>;

        /// Unpacks packed columns into an owned table.
        fn unpack_packed(packed: &PackedColumns) -> Result<UniquePtr<Table>>;

        /// Unpacks packed columns into an owned table using an explicit stream.
        fn unpack_packed_on_stream(
            packed: &PackedColumns,
            stream: usize,
        ) -> Result<UniquePtr<Table>>;

        /// Returns the host metadata as bytes.
        fn metadata_to_host(self: &PackedColumns) -> Vec<u8>;

        /// Returns the size of the host metadata in bytes.
        fn metadata_size(self: &PackedColumns) -> usize;

        /// Returns the size of the GPU data in bytes.
        fn gpu_data_size(self: &PackedColumns) -> usize;

        /// Copies the packed GPU data to a host byte vector (D2H).
        fn gpu_data_to_host(self: &PackedColumns) -> Vec<u8>;

        // -- Contiguous split --

        /// Splits a table into contiguous partitions.
        fn contiguous_split_table(
            tbl: &Table,
            splits: &[i32],
            stream: usize,
        ) -> Result<UniquePtr<PackedTableVec>>;

        /// Returns the number of partitions.
        fn size(self: &PackedTableVec) -> usize;

        /// Unpacks partition at `index` into an owned table.
        fn unpack_at(self: &PackedTableVec, index: usize) -> Result<UniquePtr<Table>>;

        /// Unpacks partition at `index` into an owned table using an explicit stream.
        fn unpack_at_on_stream(
            self: &PackedTableVec,
            index: usize,
            stream: usize,
        ) -> Result<UniquePtr<Table>>;

        /// Reconstructs packed columns from host metadata and data bytes (H2D).
        fn make_packed_from_host_parts(
            metadata: &[u8],
            gpu_data: &[u8],
            stream: usize,
        ) -> Result<UniquePtr<PackedColumns>>;
    }
}
